### PR TITLE
Free solv adaptive refit

### PR DESCRIPTION
**Pull request by jkausrelay**
- Adaptive refit free solv absolute hydration
- Use ESS<50 as criteria to generate new samples under the new ff
- Use BAR to update the ref delta f
- Recompute the loss with the new samples
- Iterate until convergence
- Results on free solv random split similar to manual run, refit, run refit loop (see https://relaytx.atlassian.net/wiki/spaces/TM/pages/198934553/WIP+Free+Solv+Refitting)

rmse (kcal/mol) on 80/20 random split
Train (this script) | 0.91 | 1.00 | 0.89 | 0.96 | 0.95
Train run 3 (manual loop) | 0.93 | 0.85 | 0.85 | 0.84 | 0.83

Test resampled (this script) | 1.24 | 0.98 | 1.39 | 1.15 | 1.55
Test run3 (manual loop) | 1.31 | 0.94 | 1.28 | 1.24 | 1.39

### CODE REVIEW

File: examples/refit_freesolv.py [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+        assert (
+            len(new_resampled_mols) == 0
+        ), f"no molecules should need to be resampled here {[get_mol_name(mol) for mol in new_resampled_mols]}"
+
+    # update the predictions to pass back to main
+    pred_dg_dict = {mol_name: a["pred_dg"] for mol_name, a in aux.items()}
+    loss_args.pred_dg_dict.update(pred_dg_dict)
+
+    # Normalize loss/grads by total number of molecules
+    return float(v / len(mols)), np.array(g / len(mols), dtype=np.float64)
```
Comment (maxentile) [inline examples/refit_freesolv.py:394]: Why is this normalization necessary?

```suggestion
    return float(v), np.array(g, dtype=np.float64)
```

File: examples/refit_freesolv.py [hunk 2]
```diff
@@ -0,0 +1,830 @@
+import jax
+
+# suppress jax warning
+jax.config.update("jax_platform_name", "cpu")
```
Comment (maxentile) [inline examples/refit_freesolv.py:4]: Actually, is this a case where GPU Jax would be permissible in `timemachine`?

**Review by maxentile — APPROVED** [review]
Sorry for my delayed review!

Some small non-blocking comments, but overall looks good!

File: examples/refit_freesolv.py [hunk 3]
```diff
@@ -0,0 +1,830 @@
...
+        description="Refit the BCC parameters for Free Solv to minimize error in predicted hydration free energies."
+    )
+    parser.add_argument("--result_path", type=str, help="path with smc results", default=".")
+    parser.add_argument("--prefactor_path", type=str, help="path to store prefactors if not empty.")
+    parser.add_argument("--n_gpus", type=int, help="number of gpus", default=get_gpu_count())
+    parser.add_argument(
+        "--ff", type=str, help="path to forcefield file or use default SMIRNOFF ff if not set", default=DEFAULT_FF
+    )
+    parser.add_argument(
+        "--ff_refit", type=str, help="path to refit forcefield - useful to test on new set of molecules", default=None
```
[the remark this answers is not in the document]
Comment (jkausrelay) [reply examples/refit_freesolv.py:98]: Please let me know if the usage still isn't clear.

File: examples/refit_freesolv.py [hunk 4]
```diff
@@ -0,0 +1,830 @@
...
+import argparse
+import copy
+import pickle
+from csv import DictWriter
+from dataclasses import dataclass
+from functools import partial
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Tuple
+
+import numpy
```
[the remark this answers is not in the document]
Comment (jkausrelay) [reply examples/refit_freesolv.py:15]: fixed

File: examples/refit_freesolv.py [hunk 5]
```diff
@@ -0,0 +1,830 @@
...
+class EndpointSamples:
+    samples_0: Samples
+    samples_1: Samples
+
+
+@dataclass
+class Prefactor:
+    prefactor: Array  # Array of prefactors to compute the electrostatic potentials
+    charge_rescale_mask: Array  # rescale mask for the ligand
+    ligand_charges: List[float]  # ligand charges used to generate the samples
```
[the remark this answers is not in the document]
Comment (jkausrelay) [reply examples/refit_freesolv.py:67]: updated

File: examples/refit_freesolv.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxentile) [reply examples/refit_freesolv.py:98]: Ahh, thanks for clarifying -- this makes sense!

File: examples/refit_freesolv.py (hunk 1 above)
Comment (jkausrelay) [reply examples/refit_freesolv.py:394]: It's to make it easier to interpret the loss independent of the number of molecules.

File: timemachine/constants.py [hunk 6]
```diff
@@ -12,3 +12,8 @@
 
 kB = unit.BOLTZMANN_CONSTANT_kB * unit.AVOGADRO_CONSTANT_NA
 DEFAULT_FF = "smirnoff_1_1_0_ccc.py"
+DEFAULT_TEMP = 300
+DEFAULT_KT = BOLTZ * DEFAULT_TEMP
+
+KCAL_TO_KJ = 4.184  # multiply to convert from kcal/mol to kJ
+KCAL_TO_KT = KCAL_TO_KJ / DEFAULT_KT
```
[the remark this answers is not in the document]
Comment (jkausrelay) [reply timemachine/constants.py:19]: Used KCAL_TO_DEFAULT_KT

File: timemachine/md/smc.py [hunk 7]
```diff
@@ -151,6 +153,12 @@ def effective_sample_size(log_weights):
     return 1 / np.sum(normed_weights ** 2)
 
 
+def jeffective_sample_size(log_weights):
```
[the remark this answers is not in the document]
Comment (jkausrelay) [reply timemachine/md/smc.py:156]: Probably, the only thing is I can't do the `assert np.isclose(np.sum(normed_weights), 1.0)` in a jax function.

[the remark this answers is not in the document]
Comment (maxentile) [reply timemachine/md/smc.py:156]: Can remove the assertion in this case I think -- but maybe this is something to revisit more generally and carefully next deboggling week...

----

(May want to remove unimportant assertions in functions we would like to `jit` compile, and/or find an idiom that handles the cases when an assertion is important AND we would like to `jit` compile the code that makes the assertion. Perhaps something along the lines of refactoring
```python
def f(x):
   ...
   assert something
   assert something_else
   return result
...
# res = jit(f)(x)  # error
res = f(x)
```
to 
```python
def f(x):
   ...
   all_assertions_passed = something and something_else
   return result, all_assertions_passed
...
result, all_assertions_passed = jit(f)(x)
assert all_assertions_passed
```
, so that the assertion conditions are still computed, but the assertion is moved outside of the `jit` compiled function...)

File: examples/refit_freesolv.py (hunk 1 above)
Comment (maxentile) [reply examples/refit_freesolv.py:394]: I think it would be preferable to move the normalization outside of this function.

(So that
(1) `loss` has consistent behavior with `_loss_serial`, `loss_serial`, `loss_parallel`, which all return `sum` instead of `mean`,
(2) `loss(..., dataset_a) + loss(..., dataset_b) == loss(..., combine(dataset_a, dataset_b))`
(3) we don't risk accidentally dividing by `len(mols)` twice
(4) we won't need to remember to multiply by `len(mols)` if we want to add parameter regularization (`regularize(params) + loss(params, data)`, where `loss` is expected to be size-extensive with `len(data)`.)
